### DISCUSSION TITLE
Clearer Names / module hierarchy

### DISCUSSION THREAD

**Discussion by epage** [winnow-rs/winnow#95] [Q&A]
Some of the original nom naming is confusing for new people.  How can we name things so a casual programmer (not steeped in specific parsing backgrounds) can walk up and use this crate while being technically clear to help as people need to do more advanced things?

e.g.
- [confusion over `tag`](https://hachyderm.io/@pearfalse@tech.lgbt/109767208540750093)
- [hard to search all the modules](https://github.com/rust-bakery/nom/issues/1506)

I've started a [spreadsheet](https://docs.google.com/spreadsheets/d/199iOQ5Xi6hcw5EgY0zTsCz7ls9BA9vhJxWq9bzRUa2g/edit?usp=sharing) of the current names, with comparisons to some other parsers.  I'd also be open to comparing with more parsers (not just Rust ones) if there are good points of inspiration.

**Comment by epage** [comment]
See also
- rust-bakery/nom#590
- rust-bakery/nom#968
- rust-bakery/nom#1468

**Comment by epage** [comment]
Running list of whats been done
- #4 Move `Map`, etc to `combinator::Map`, etc
- #28 Move `*::complete::*` / `*::streaming::*` into `*::*` by auto-detecting complete/streaming
- #38 Moved `number::float` to `character::f32` and `number::double` to `character::f64`
- #40 
  - Renamed `combinator::map_res` with `Parser::map_res`
  - Renamed `combinator::map_opt` with `Parser::map_opt`
  - Renamed `combinator::complete` with `Parser::complete`
  - Renamed `combinator::verify` with `Parser::verify`
  - Renamed `combinator::value` with `Parser::value`
  - Renamed `combinator::recognize` with `Parser::recognize`
  - Renamed `combinator::consumed` with `Parser::with_recognized`
  - Renamed  `error::context` with `Parser::context`
  - Renamed `error::dbg_dmp` with `Parser::dbg_err`
- #44 Generalized `character::*::anychar` into `bytes::any`
- #48 Replace `Parser::into` with `Parser::output_into` and `Parser::err_into` (`Parser::into` conflicted with `Into::into`)
- #52 `pub use bits::*` in root namespace
- #13 + #73 Move `ErrorConvert` to `error::ErrorConvert`
- #76 Renamed `Parser::parse` to `Parsr::parse_next` to clarify the incremental nature
- #105 misc renames related to input that don't affect most users
- #108 moved `bytes::escaped` into `character::escaped`
- #117
  - Renamed `Err { Incomplete, Error, Failure }` to `ErrMode { Incomplete, Backtrack, Cut }`
  - Renamed `cut` to `cut_err`
  - Copied `IResult` into `error`
  - Moved `Needed` and `ErrMode` into `error`
  - Moved `error::Context` and `err::DbgDump` into `combinator`
- #129
  - Merge `character::f32` and `character::f64` into `character::float`
  - Generalize `number::hex_u32` into `character::hex_uint`
  - Merge `character::u8`, `character::u16`, etc into `character::dec_uint` and `character::dec_int`
- #127 Deprecate `many0_count`, etc in favor of `many0` with new inferred output parameter
- #132 Renamed combinators that were missing the `0` suffix
- #140 `separated_list0` to `separated0` and `separated_list1` to `separated1` to open the room for other `separated` functions
- #13 + #142 Moved input types to `stream`
- #142 
  - `Streaming` to `Partial` to open the door for changing some uses of "input" to "stream" which both makes it clearer that its incremental, like in #76, and in case we change how we use it in #72
  - misc renames related to input that don't affect most users
- #145 Renamed `Parser::map_opt` to `Parser::verify_map` as its a `verify` + `map`, paralleling `filter` + `map` is `filter_map`
- #239 Handles module moves
- #240

**Comment by epage** [comment]
`nom` tries to use the module heiarchy as a ToC for the different combinators (e.g. `sequence`) with a catch-all (`combinators`).

I think I want to instead switch it around based on roles:
- `combinators` has all the role-agnostic combinators (`combinators`, `sequence`, `bytes`, etc)
- `str` is the new name for `character`
- `binary` is the new name for `number`
- `binary::bits` is the new name for `bits` (since they overlap with other combinators like `tag`, this keeps them scoped with wildcard imports)

This means the error types in #103 will likely be role specific (`BStrError` and `BinError`).  That, combined with the above, makes me wonder if we should have role-based `IResult` aliases that default the error to their role's specific one.  That then makes me wonder if we should have role-specific preludes.

**Reply by epage** [reply]
#239 implements this

**Comment by vwkd** [comment]
It seems the naming currently uses a mix of verbs, nouns, adjectives, and other parts of speech.

For example, currently the name of the `separated` combinator is an adjective while the name of the `repeat` combinator is a verb, which previously was yet another part of speech (`many`).

It would be good to specify a naming convention.

Some more examples of current names:

verbs
- `repeat`
- `fill`
- `fold_repeat`
- `fail`
- `map`
- `peek`
- `parse`
- etc.

adjectives
- `separated`
- `escaped`
- `terminated`
- `delimited`
- `preceded`
- etc.

nouns
- `alt`(ernative)
- `success`
- `digit`
- `newline`
- etc.

other (?)
- `any`
- `one_of`
- `none_of`

**Reply by epage** [reply]
I expect we'll at least have a mix of nouns and verbs or adjectives considering the different role different parsers may take.  Whether we do both verbs and adjectives or lean on one is yet to be decided.

I did an incomplete survey of other parsers.  Might be worth looking at those for ideas

> I've started a [spreadsheet](https://docs.google.com/spreadsheets/d/199iOQ5Xi6hcw5EgY0zTsCz7ls9BA9vhJxWq9bzRUa2g/edit?usp=sharing) of the current names, with comparisons to some other parsers. I'd also be open to comparing with more parsers (not just Rust ones) if there are good points of inspiration.
